File: lab2-functionalPrimer/javascript/func_as_arg.js
// Create list of ints from a to b (inclusive)
function arrFunc(a, b) {
    let arr = [];
    for (let i = a; i <= b; i++) {
        arr.push(i);
    }
    return arr;
}

// applicatorFunc: takes a function, start, end, and choice ('s' or 'a')
function applicatorFunc(inpFunc, a, b, choice) {
    const arr = inpFunc(a, b);
    const sum = arr.reduce((accumulator, currentValue) => accumulator + currentValue, 0);

    if (choice === 's') {
        return sum;
    } else if (choice === 'a') {
        return sum / arr.length;
    } else {
        throw new Error("Invalid choice! Use 's' for sum or 'a' for average.");
    }
}

// ===== User Input Section =====
const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

readline.question("Enter start of range (a): ", aStr => {
    readline.question("Enter end of range (b): ", bStr => {
        readline.question("Enter 's' for sum or 'a' for average: ", choiceStr => {
            const a = parseInt(aStr);
            const b = parseInt(bStr);
            const choice = choiceStr.trim().toLowerCase();

            try {
                const result = applicatorFunc(arrFunc, a, b, choice);
                if (choice === 's') {
                    console.log(`Sum = ${result}`);
                } else {
                    console.log(`Average = ${result}`);
                }
            } catch (err) {
                console.error(err.message);
            }

            readline.close();
        });
    });
});

File: lab2-functionalPrimer/python3/func_as_arg.py
def listFunc(a, b):
    # Create list of ints from a to b (inclusive)
    return [i for i in range(a, b + 1)]

def applicatorFunc(inpFunc, s, a, b):
    if s == 's':
        return sum(inpFunc(a, b))
    elif s == 'a':
        return sum(inpFunc(a, b)) / (b - a + 1)
    else:
        raise ValueError("Invalid choice! Use 's' for sum or 'a' for average.")

# Get user input and convert to integers
a = int(input("Enter start of range (a): "))
b = int(input("Enter end of range (b): "))
choice = input("Enter 's' for sum or 'a' for average: ").strip().lower()

# Call and print result
result = applicatorFunc(listFunc, choice, a, b)
print("Result =", result)